From 5972387dec85bbefc3bd3ace1db4875abd748c4c Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 8 Dec 2025 17:55:27 +0800
Subject: [PATCH 250/254] mtk: hostapd: add 'allocated_links' to track
 allocated link on an AP MLD

The hapd->mld_link_id is allocated in hostapd_bss_alloc_link_id(),
while hapd->mld->valid_links is updated in hostapd_mld_add_link(),
which is called by hostapd_setup_bss().

The problem occurs when the setup is done by directly calling hostapd
with configuration files as input parameters
(e.g., hostapd hostapd.conf1 hostapd.conf2).

The hapd->mld_link_id is allocated first for each hapd of the AP MLD,
then the BSS setup process is called. Since 'hapd->mld->valid_links'
is always 0 in this case, the hapd->mld_link_id will be 0 for each
hapd, causing the setup to fail.

Add a new variable "allocated_links" to avoid the problem.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/hostapd.c | 4 +++-
 src/ap/hostapd.h | 1 +
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index d04c8cba1..740ef916a 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -3417,8 +3417,9 @@ static void hostapd_bss_alloc_link_id(struct hostapd_data *hapd)
 		int i;
 
 		for (i = 0; i < MAX_NUM_MLD_LINKS; i++) {
-			if ((hapd->mld->valid_links & BIT(i)) == 0) {
+			if ((hapd->mld->allocated_links & BIT(i)) == 0) {
 				hapd->mld_link_id = i;
+				hapd->mld->allocated_links |= BIT(i);
 				break;
 			}
 		}
@@ -5923,6 +5924,7 @@ int hostapd_mld_remove_link(struct hostapd_data *hapd)
 	dl_list_del(&hapd->link);
 	mld->num_links--;
 	mld->valid_links &= ~BIT(hapd->mld_link_id);
+	mld->allocated_links &= ~BIT(hapd->mld_link_id);
 
 	wpa_printf(MSG_DEBUG, "AP MLD %s: Link ID %d removed. num_links: %d",
 		   mld->name, hapd->mld_link_id, mld->num_links);
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index 8b65b932e..4c140675b 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -605,6 +605,7 @@ struct hostapd_mld {
 	u8 refcount;
 	bool started;
 	u16 link_reconf_in_progress;
+	u16 allocated_links;
 	u16 valid_links;
 	u16 removed_links;
 	u8 eml_disable;
-- 
2.45.2

